### PR TITLE
[TypeScript] Implement Dependency Injection in TypeScript bots

### PR DESCRIPTION
**Pull request by Batta32**
### Purpose
*What is the context of this pull request? Why is it being done?*

### Changes
*Are there any changes that need to be called out as significant or particularly difficult to grasp? (Include illustrative screenshots for context if applicable.)*

### Tests
*Is this covered by existing tests or new ones? If no, why not?*

### Feature Plan
*Are there any remaining steps or dependencies before this issue can be fully resolved? If so, describe and link to any relevant pull requests or issues.*

#### General
- [ ] I have commented my code, particularly in hard-to-understand areas	
- [ ] I have added or updated the appropriate tests	
- [ ] I have updated related documentation

### CODE REVIEW

File: sdk/typescript/libraries/bot-solutions/src/responses/localeTemplateManager.ts [hunk 1]
```diff
@@ -25,7 +25,6 @@ export class LocaleTemplateManager extends MultiLanguageLG {
         if (fallbackLocale !== undefined && fallbackLocale.trim().length === 0) {
             throw new Error(`'fallbackLocale' shouldn't be empty string. If you don't want to set it, please set it to undefined.`);
         }
-
```
Comment (Batta32) [inline sdk/typescript/libraries/bot-solutions/src/responses/localeTemplateManager.ts:28]: I'd like to remove this.

File: templates/typescript/samples/sample-assistant/src/adapters/defaultAdapter.ts [hunk 2]
```diff
@@ -23,16 +23,18 @@ import {
     FeedbackOptions} from 'bot-solutions';
 import { TelemetryInitializerMiddleware } from 'botbuilder-applicationinsights';
 import { IBotSettings } from '../services/botSettings.js';
+import { inject } from 'inversify';
+import { TYPES } from '../types/constants.js';
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/adapters/defaultAdapter.ts:27]:
```suggestion
import { TYPES } from '../types/constants';
```

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 3]
```diff
@@ -95,9 +101,11 @@ export class MainDialog extends ComponentDialog {
...
         this.addDialog(this.switchSkillDialog);
         // Register skill dialogs
-        skillDialogs.forEach((skillDialog: SkillDialog): void => {
-            this.addDialog(skillDialog);
-        });
+        if (skillDialogs) {
+            skillDialogs.forEach((skillDialog: SkillDialog): void => {
+                this.addDialog(skillDialog);
+            });
+        }
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts:108]: Is this okay? Is this related to `skillDialogs` being `optional`?
I don't think that this is correct.

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 4]
```diff
@@ -393,7 +401,11 @@ export class MainDialog extends ComponentDialog {
         } else {
             
             return await stepContext.next();
-        }
+            
+        } 
+            
+        return await stepContext.next();
+
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts:408]: Why this?

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 5]
```diff
@@ -464,7 +476,7 @@ export class MainDialog extends ComponentDialog {
 
         return await next();
     }
-
+
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts:479]: Why this?

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 6]
```diff
@@ -474,37 +486,32 @@ export class MainDialog extends ComponentDialog {
         }
 
         return true;
-    }
+    }
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts:489]: Why this?

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 7]
```diff
@@ -474,37 +486,32 @@ export class MainDialog extends ComponentDialog {
...
-     * @param stepContext - current dialog context.
-     * @param dispatchIntent - Intent that Dispatch thinks should be invoked.
-     * @param dispatchScore - Confidence score for intent.
-     * @param threshold - user provided threshold between 0.0 and 1.0, if above this threshold do NOT show chitchat.
-     * @returns A bool indicating if we should invoke the personality dialog.
-     */
-    private shouldBeginChitChatDialog(stepContext: WaterfallStepContext, dispatchIntent: string, dispatchScore: number, threshold = 0.5): boolean {
-        if (threshold < 0.0 || threshold > 1.0) {
-            throw new Error(`The argument ${ threshold } is out of range`);
+    private shouldBeginChitChatDialog(stepContext: WaterfallStepContext, dispatchIntent: string, dispatchScore: number, threshold = 0.5): boolean{
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts:491]: Is this okay? Should we remove the comments?

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 8]
```diff
@@ -474,37 +486,32 @@ export class MainDialog extends ComponentDialog {
...
             }
-        } else if (dispatchScore < threshold) {
+        }
+
+        else if (dispatchScore < threshold){
             return true;
         }
 
         return false;
+
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts:515]: Can we confirm this to the current next branch? I don't think that this is related to Dependency Injection

File: templates/typescript/samples/sample-assistant/tsconfig.json [hunk 9]
```diff
@@ -1,9 +1,9 @@
 {
   "compilerOptions": {
     /* Basic Options */
-    "target": "ES2017",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
+    "target": "ES2016",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/tsconfig.json:4]: I'd like to confirm all these changes to `tsconfig`, are really weird to change the `target`, and the `lib`.
Can we check this?

File: templates/typescript/samples/sample-skill/src/dialogs/skillDialogBase.ts [hunk 10]
```diff
@@ -31,12 +34,12 @@ export class SkillDialogBase extends ComponentDialog {
     protected templateManager: LocaleTemplateManager;
 
     public constructor(
-        dialogId: string,
-        settings: Partial<IBotSettings>,
-        services: BotServices,
-        stateAccessor: StatePropertyAccessor<SkillState>,
-        telemetryClient: BotTelemetryClient,
-        templateManager: LocaleTemplateManager
+        @unmanaged() dialogId: string,
```
Comment (Batta32) [inline templates/typescript/samples/sample-skill/src/dialogs/skillDialogBase.ts:37]: Is this okay to be `@unmanaged`?

File: templates/typescript/samples/sample-skill/tsconfig.json [hunk 11]
```diff
@@ -3,7 +3,7 @@
     /* Basic Options */
     "target": "ES2017",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
     "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
-    // "lib": [],                             /* Specify library files to be included in the compilation. */
```
Comment (Batta32) [inline templates/typescript/samples/sample-skill/tsconfig.json:6]: Same comments of the VA's `tsconfig.json`

File: templates/typescript/samples/sample-assistant/tsconfig.json (hunk 9 above)
Comment (LCarbonellRod) [reply templates/typescript/samples/sample-assistant/tsconfig.json:4]: I rollbacked this change, and tested and worked fine with 2017.

File: templates/typescript/samples/sample-skill/tsconfig.json (hunk 11 above)
Comment (LCarbonellRod) [reply templates/typescript/samples/sample-skill/tsconfig.json:6]: This change is needed as the [inversify's doc](http://inversify.io/#step_1) says.

File: templates/typescript/samples/sample-skill/src/dialogs/skillDialogBase.ts (hunk 10 above)
Comment (LCarbonellRod) [reply templates/typescript/samples/sample-skill/src/dialogs/skillDialogBase.ts:37]: We need it as the classes which inherit from this class don't use the `dialogId` property. You can check more info about this in the [documentation](https://varie.io/docs/latest/dependency-injection#extending-a-injectable-class).

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts (hunk 8 above)
Comment (LCarbonellRod) [reply templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts:515]: I'll rollback the changes which don't have to do with dependency injection.

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts (hunk 7 above)
Comment (LCarbonellRod) [reply templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts:491]: I'll rolback this.

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts (hunk 6 above)
Comment (LCarbonellRod) [reply templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts:489]: I'll rolback this.

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts (hunk 5 above)
Comment (LCarbonellRod) [reply templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts:479]: I'll rolback this.

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts (hunk 4 above)
Comment (LCarbonellRod) [reply templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts:408]: I'll rolback this.

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts (hunk 3 above)
Comment (LCarbonellRod) [reply templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts:108]: I'll rolback this.

File: templates/typescript/samples/sample-skill/src/dialogs/skillDialogBase.ts (hunk 10 above)
Comment (Batta32) [reply templates/typescript/samples/sample-skill/src/dialogs/skillDialogBase.ts]: Ok, I understood

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 12]
```diff
@@ -57,15 +59,15 @@ export class MainDialog extends ComponentDialog {
...
-        activeSkillProperty: StatePropertyAccessor<BotFrameworkSkill>
+        @inject(TYPES.BotServices) services: BotServices,
+        @inject(TYPES.LocaleTemplateManager) templateManager: LocaleTemplateManager,
+        @inject(TYPES.IUserProfileState) userProfileState: StatePropertyAccessor<IUserProfileState>,
+        @inject(TYPES.Activity) previousResponseAccessor: StatePropertyAccessor<Partial<Activity>[]>,
+        @inject(TYPES.OnboardingDialog) onBoardingDialog: OnboardingDialog,
+        @inject(TYPES.SwitchSkillDialog) switchSkillDialog: SwitchSkillDialog,
+        @inject(TYPES.SkillDialogs) @optional() skillDialogs: SkillDialog[],
+        @inject(TYPES.SkillsConfiguration) @optional() skillsConfig: SkillsConfiguration,
+        @inject(TYPES.BotFrameworkSkill) @optional() activeSkillProperty: StatePropertyAccessor<BotFrameworkSkill>
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts:70]: Can you explain why these should be `@optional`?

File: templates/typescript/samples/sample-assistant/src/inversify.config.ts [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+
+function getTelemetryClient(settings: Partial<IBotSettings>): BotTelemetryClient {
+    if (settings !== undefined && settings.appInsights !== undefined && settings.appInsights.instrumentationKey !== undefined) {
+        const instrumentationKey: string = settings.appInsights.instrumentationKey;
+
+        return new ApplicationInsightsTelemetryClient(instrumentationKey);
+    }
+
+    return new NullTelemetryClient();
+}
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/inversify.config.ts:246]:
```suggestion
function getTelemetryClient(settings: Partial<IBotSettings>): BotTelemetryClient {
    if (settings !== undefined && settings.appInsights !== undefined && settings.appInsights.instrumentationKey !== undefined) {
        const instrumentationKey: string = settings.appInsights.instrumentationKey;

        return new ApplicationInsightsTelemetryClient(instrumentationKey);
    }

    return new NullTelemetryClient();
}

export default container;
```

nit: because the last thing to do in this file is to export the `container`

File: templates/typescript/samples/sample-assistant/tsconfig.json [hunk 14]
```diff
@@ -3,7 +3,7 @@
     /* Basic Options */
     "target": "ES2017",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
     "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
-    // "lib": [],                             /* Specify library files to be included in the compilation. */
+    "lib": ["es6"],                           /* Specify library files to be included in the compilation. */
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/tsconfig.json:6]: As the target is `ES2017`, we can use lib related to that `target`? I don't know if this exists.

File: templates/typescript/samples/sample-skill/src/inversify.config.ts [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+
+function getTelemetryClient(settings: Partial<IBotSettings>): BotTelemetryClient {
+    if (settings !== undefined && settings.appInsights !== undefined && settings.appInsights.instrumentationKey !== undefined) {
+        const instrumentationKey: string = settings.appInsights.instrumentationKey;
+
+        return new ApplicationInsightsTelemetryClient(instrumentationKey);
+    }
+
+    return new NullTelemetryClient();
+}
```
Comment (Batta32) [inline templates/typescript/samples/sample-skill/src/inversify.config.ts:169]:
```suggestion
function getTelemetryClient(settings: Partial<IBotSettings>): BotTelemetryClient {
    if (settings !== undefined && settings.appInsights !== undefined && settings.appInsights.instrumentationKey !== undefined) {
        const instrumentationKey: string = settings.appInsights.instrumentationKey;

        return new ApplicationInsightsTelemetryClient(instrumentationKey);
    }

    return new NullTelemetryClient();
}

export default container;
```

File: templates/typescript/samples/sample-skill/tsconfig.json (hunk 11 above)
Comment (Batta32) [reply templates/typescript/samples/sample-skill/tsconfig.json:6]: Same comment of the VA

File: templates/typescript/samples/sample-assistant/tsconfig.json (hunk 14 above)
Comment (LCarbonellRod) [reply templates/typescript/samples/sample-assistant/tsconfig.json:6]: It seems that the [inversify documentation](http://inversify.io/#step_1) needs to use `lib` with this value.

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts (hunk 12 above)
Comment (LCarbonellRod) [reply templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts:70]: As shown below, the SkillDialog and SkillsConfiguration classes are inside an `if else` statement, if the VA doesn't have any skill registered, the mentioned classes are not registered when the `inversify.config.ts` file is requested and the Bot can't start. Regarding the `BotFrameworkSkill`, I will rollback this change.

_registering SkillDialog and SkillsConfiguration classes in the inversify.config.ts file_
image

**Comment by LCarbonellRod** [comment]
> 1. Can you run the linter here?
> 2. Can you update the `package-lock.json` of each sample?
> 
> After these changes, we should migrate these to the generator (templates) and check the end-to-end using the VA/Skill of the generator.

I ran linter and I only got warnings.

_Virtual Assistant warnings_
image

_Skill warnings_
image

**Comment by Batta32** [comment]
As you shown, there are some warnings that are potentially fixable with `--fix` option. It would be great to fix them and update the templates with that.

As a final step, you can validate the end-to-end using the generator. Don't forget to check the generator's test!

**Comment by Batta32** [comment]
LGTM! Can you clean this branch?
